web/modules/custom/brms_sor/src/Service/FmsSorApi.php

```diff
@@ -0,0 +1,252 @@
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelFactoryInterface
+   */
+  protected $logger;
+
+  /**
+   * The JSON serializer service.
+   *
+   * @var \Drupal\Component\Serialization\Json
+   */
+  protected $json;
+
+  /**
+   * The cache backend.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cache;
+
+  /**
+   * FMS API auth URL.
+   *
+   * @var string
+   */
+  protected $authUrl;
+
+  /**
+   * FMS API base URL.
+   *
+   * @var string
+   */
+  protected $baseUrl;
+
+  /**
+   * FMS API client id.
+   *
+   * @var string
+   */
+  protected $clientId;
+
+  /**
+   * Secret of the DnB client.
+   *
+   * @var string
+   */
+  protected $clientSecret;
+
+  /**
+   * Constructs a FmsSorApi object.
+   *
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The GuzzleHttp Client.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   The logger factory.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cache backend.
+   * @param \Drupal\key\KeyRepositoryInterface $key_repository
+   *   The key repository.
+   * @param \Drupal\Component\Serialization\Json $json_service
+   *   The JSON serializer service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   */
+  public function __construct(
+    protected ClientInterface $http_client,
+    protected LoggerChannelFactoryInterface $logger_factory,
+    protected CacheBackendInterface $cache_backend,
+    protected KeyRepositoryInterface $key_repository,
+    protected Json $json_service,
+    protected TimeInterface $time,
+  ) {
+    $this->httpClient = $http_client;
+    $this->logger = $logger_factory;
+    $this->cache = $cache_backend;
+    $this->json = $json_service;
+    $settings = Settings::get('brms_sor');
+    $this->authUrl = $settings['fms_api_auth_url'];
+    $this->baseUrl = $settings['fms_api_base_url'];
+    $this->clientId = $settings['fms_api_client_id'];
+    $this->clientSecret = $key_repository->getKey('fms_api')->getKeyValue();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessToken(bool $use_cached_token = TRUE): string {
+    if ($use_cached_token) {
+      $cache_access_token = $this->cache->get('brms_sor_fms_access_token');
+      if ($cache_access_token) {
+        return $cache_access_token->data;
+      }
+    }
+
+    try {
+      $response = $this->httpClient->request('POST', $this->authUrl, [
+        'headers' => [
+          'Content-Type' => 'application/x-www-form-urlencoded',
+        ],
+        'form_params' => [
+          'grant_type' => 'client_credentials',
+          'client_id' => $this->clientId,
+          'client_secret' => $this->clientSecret,
+        ],
+      ]);
+
+      $data = $this->json->decode($response->getBody()->getContents());
+      $this->cache->set('brms_sor_fms_access_token', $data['access_token'], $this->time->getRequestTime() + $data['expires_in']);
```

Comment:
We control the expires_in and the expiration date, so I would say we don't need to cache this, just see if the token can be renewed based on timestamps

Reply:
I'm still caching the token because I don't want to call the API to generate a new once for each get and patch. I added logic that checks if the token is about to expire, and if it's within a minute of expiring a new one will be generated.

Reply:
Let's use tempstore as discussed